// fireblocks-sdk/dist/api-token-provider.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiTokenProvider = void 0;
const jsonwebtoken_1 = __importDefault(require("jsonwebtoken"));
const crypto_1 = __importDefault(require("crypto"));
const uuid_1 = require("uuid");
class ApiTokenProvider {
    constructor(privateKey, apiKey) {
        this.privateKey = privateKey;
        this.apiKey = apiKey;
    }
    signJwt(path, bodyJson) {
        const token = jsonwebtoken_1.default.sign({
            uri: path,
            nonce: uuid_1.v4(),
            iat: Math.floor(Date.now() / 1000),
            exp: Math.floor(Date.now() / 1000) + 55,
            sub: this.apiKey,
            bodyHash: crypto_1.default.createHash("sha256").update(JSON.stringify(bodyJson || "")).digest().toString("hex")
        }, this.privateKey, { algorithm: "RS256" });
        return token;
    }
    getApiKey() {
        return this.apiKey;
    }
}
exports.ApiTokenProvider = ApiTokenProvider;
