// fireblocks-sdk/dist/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransferTicketTermStatus = exports.TransferTicketStatus = exports.TransactionOperation = exports.VirtualType = exports.PeerType = exports.TransactionStatus = exports.TransactionOrder = exports.NetworkDestType = exports.Scheme = exports.FeeLevel = exports.SigningAlgorithm = exports.VaultAssetActivationStatus = exports.VirtualAffiliation = void 0;
var VirtualAffiliation;
(function (VirtualAffiliation) {
    VirtualAffiliation["OFF_EXCHANGE"] = "OFF_EXCHANGE";
    VirtualAffiliation["DEFAULT"] = "DEFAULT";
})(VirtualAffiliation = exports.VirtualAffiliation || (exports.VirtualAffiliation = {}));
var VaultAssetActivationStatus;
(function (VaultAssetActivationStatus) {
    VaultAssetActivationStatus["PENDING_ACTIVATION"] = "PENDING_ACTIVATION";
    VaultAssetActivationStatus["ACTIVATION_FAILED"] = "ACTIVATION_FAILED";
    VaultAssetActivationStatus["READY"] = "READY";
})(VaultAssetActivationStatus = exports.VaultAssetActivationStatus || (exports.VaultAssetActivationStatus = {}));
var SigningAlgorithm;
(function (SigningAlgorithm) {
    SigningAlgorithm["MPC_ECDSA_SECP256K1"] = "MPC_ECDSA_SECP256K1";
    SigningAlgorithm["MPC_ECDSA_SECP256R1"] = "MPC_ECDSA_SECP256R1";
    SigningAlgorithm["MPC_EDDSA_ED25519"] = "MPC_EDDSA_ED25519";
})(SigningAlgorithm = exports.SigningAlgorithm || (exports.SigningAlgorithm = {}));
var FeeLevel;
(function (FeeLevel) {
    FeeLevel["HIGH"] = "HIGH";
    FeeLevel["MEDIUM"] = "MEDIUM";
    FeeLevel["LOW"] = "LOW";
})(FeeLevel = exports.FeeLevel || (exports.FeeLevel = {}));
var Scheme;
(function (Scheme) {
    Scheme["AUTO"] = "AUTO";
    Scheme["DEFAULT"] = "DEFAULT";
    Scheme["CUSTOM"] = "CUSTOM";
})(Scheme = exports.Scheme || (exports.Scheme = {}));
var NetworkDestType;
(function (NetworkDestType) {
    NetworkDestType["VAULT_ACCOUNT"] = "VAULT";
    NetworkDestType["UNMANAGED_WALLET"] = "UNMANAGED";
    NetworkDestType["EXCHANGE_ACCOUNT"] = "EXCHANGE";
    NetworkDestType["FIAT_ACCOUNT"] = "FIAT_ACCOUNT";
})(NetworkDestType = exports.NetworkDestType || (exports.NetworkDestType = {}));
var TransactionOrder;
(function (TransactionOrder) {
    TransactionOrder["CREATED_AT"] = "createdAt";
    TransactionOrder["LAST_UPDATED"] = "lastUpdated";
})(TransactionOrder = exports.TransactionOrder || (exports.TransactionOrder = {}));
var TransactionStatus;
(function (TransactionStatus) {
    TransactionStatus["SUBMITTED"] = "SUBMITTED";
    TransactionStatus["QUEUED"] = "QUEUED";
    TransactionStatus["PENDING_SIGNATURE"] = "PENDING_SIGNATURE";
    TransactionStatus["PENDING_AUTHORIZATION"] = "PENDING_AUTHORIZATION";
    TransactionStatus["PENDING_3RD_PARTY_MANUAL_APPROVAL"] = "PENDING_3RD_PARTY_MANUAL_APPROVAL";
    TransactionStatus["PENDING_3RD_PARTY"] = "PENDING_3RD_PARTY";
    /**
     * @deprecated
     */
    TransactionStatus["PENDING"] = "PENDING";
    TransactionStatus["BROADCASTING"] = "BROADCASTING";
    TransactionStatus["CONFIRMING"] = "CONFIRMING";
    /**
     * @deprecated Replaced by "COMPLETED"
     */
    TransactionStatus["CONFIRMED"] = "CONFIRMED";
    TransactionStatus["COMPLETED"] = "COMPLETED";
    TransactionStatus["PENDING_AML_SCREENING"] = "PENDING_AML_SCREENING";
    TransactionStatus["PARTIALLY_COMPLETED"] = "PARTIALLY_COMPLETED";
    TransactionStatus["CANCELLING"] = "CANCELLING";
    TransactionStatus["CANCELLED"] = "CANCELLED";
    TransactionStatus["REJECTED"] = "REJECTED";
    TransactionStatus["FAILED"] = "FAILED";
    TransactionStatus["TIMEOUT"] = "TIMEOUT";
    TransactionStatus["BLOCKED"] = "BLOCKED";
})(TransactionStatus = exports.TransactionStatus || (exports.TransactionStatus = {}));
var PeerType;
(function (PeerType) {
    PeerType["VAULT_ACCOUNT"] = "VAULT_ACCOUNT";
    PeerType["EXCHANGE_ACCOUNT"] = "EXCHANGE_ACCOUNT";
    PeerType["INTERNAL_WALLET"] = "INTERNAL_WALLET";
    PeerType["EXTERNAL_WALLET"] = "EXTERNAL_WALLET";
    PeerType["UNKNOWN"] = "UNKNOWN";
    PeerType["NETWORK_CONNECTION"] = "NETWORK_CONNECTION";
    PeerType["FIAT_ACCOUNT"] = "FIAT_ACCOUNT";
    PeerType["COMPOUND"] = "COMPOUND";
    PeerType["ONE_TIME_ADDRESS"] = "ONE_TIME_ADDRESS";
    PeerType["OEC_PARTNER"] = "OEC_PARTNER";
})(PeerType = exports.PeerType || (exports.PeerType = {}));
var VirtualType;
(function (VirtualType) {
    VirtualType["OFF_EXCHANGE"] = "OFF_EXCHANGE";
    VirtualType["DEFAULT"] = "DEFAULT";
    VirtualType["OEC_FEE_BANK"] = "OEC_FEE_BANK";
})(VirtualType = exports.VirtualType || (exports.VirtualType = {}));
var TransactionOperation;
(function (TransactionOperation) {
    TransactionOperation["TRANSFER"] = "TRANSFER";
    TransactionOperation["MINT"] = "MINT";
    TransactionOperation["BURN"] = "BURN";
    TransactionOperation["SUPPLY_TO_COMPOUND"] = "SUPPLY_TO_COMPOUND";
    TransactionOperation["REDEEM_FROM_COMPOUND"] = "REDEEM_FROM_COMPOUND";
    TransactionOperation["RAW"] = "RAW";
    TransactionOperation["CONTRACT_CALL"] = "CONTRACT_CALL";
    TransactionOperation["TYPED_MESSAGE"] = "TYPED_MESSAGE";
})(TransactionOperation = exports.TransactionOperation || (exports.TransactionOperation = {}));
var TransferTicketStatus;
(function (TransferTicketStatus) {
    TransferTicketStatus["OPEN"] = "OPEN";
    TransferTicketStatus["PARTIALLY_FULFILLED"] = "PARTIALLY_FULFILLED";
    TransferTicketStatus["FULFILLED"] = "FULFILLED";
    TransferTicketStatus["FAILED"] = "FAILED";
    TransferTicketStatus["CANCELED"] = "CANCELED";
})(TransferTicketStatus = exports.TransferTicketStatus || (exports.TransferTicketStatus = {}));
var TransferTicketTermStatus;
(function (TransferTicketTermStatus) {
    TransferTicketTermStatus["OPEN"] = "OPEN";
    TransferTicketTermStatus["FULFILLED"] = "FULFILLED";
})(TransferTicketTermStatus = exports.TransferTicketTermStatus || (exports.TransferTicketTermStatus = {}));
